uamqp/aio/_client_async.py

```diff
@@ -209,6 +287,176 @@ async def do_work_async(self, **kwargs):
+
+    async def _on_send_complete_async(self, message_delivery, reason, state):
+        # TODO: check whether the callback would be called in case of message expiry or link going down
+        #  and if so handle the state in the callback
+        message_delivery.reason = reason
+        if reason == LinkDeliverySettleReason.DispositionReceived:
+            if state and SEND_DISPOSITION_ACCEPT in state:
+                message_delivery.state = MessageDeliveryState.Ok
+            else:
+                try:
+                    error_info = state[SEND_DISPOSITION_REJECT]
+                    self._process_send_error(
+                        message_delivery,
+                        condition=error_info[0][0],
+                        description=error_info[0][1],
+                        info=error_info[0][2]
+                    )
+                except TypeError:
+                    self._process_send_error(
+                        message_delivery,
+                        condition=ErrorCondition.UnknownError
+                    )
+        elif reason == LinkDeliverySettleReason.Settled:
+            message_delivery.state = MessageDeliveryState.Ok
+        elif reason == LinkDeliverySettleReason.Timeout:
+            message_delivery.state = MessageDeliveryState.Timeout
+            message_delivery.error = TimeoutError("Sending message timed out.")
+        else:
+            # NotDelivered and other unknown errors
+            self._process_send_error(
+                message_delivery,
+                condition=ErrorCondition.UnknownError
+            )
+
+    async def _send_message_impl_async(self, message, **kwargs):
+        timeout = kwargs.pop("timeout", 0)
+        expire_time = (time.time() + timeout) if timeout else None
+        await self.open_async()
+        message_delivery = _MessageDelivery(
+            message,
+            MessageDeliveryState.WaitingToBeSent,
+            expire_time
+        )
+
+        while not await self.client_ready_async():
+            await asyncio.sleep(0.05)
+
+        await self._transfer_message_async(message_delivery, timeout)
+
+        running = True
+        while running and message_delivery.state not in MESSAGE_DELIVERY_DONE_STATES:
+            running = await self.do_work_async()
+            if message_delivery.expiry and time.time() > message_delivery.expiry:
+                await self._on_send_complete_async(message_delivery, LinkDeliverySettleReason.Timeout, None)
+
+        if message_delivery.state in (
+            MessageDeliveryState.Error,
+            MessageDeliveryState.Cancelled,
+            MessageDeliveryState.Timeout
+        ):
+            try:
+                raise message_delivery.error
+            except TypeError:
+                # This is a default handler
+                raise MessageException(condition=ErrorCondition.UnknownError, description="Send failed.")
+
+    async def send_message_async(self, message, **kwargs):
```

Comment:
We should get rid of all the `_async` suffixes

Reply:
I think we could resolve the issue at the same time when we resolve the client hierarchy issue.

Reply:
using the same issue for tracking: https://github.com/Azure/azure-sdk-for-python/issues/22972